File: Python/leetcode_463.py
from typing import List


def island_perimeter(grid: List[List[int]]) -> int:
    perimeter = 0
    rows, cols = len(grid), len(grid[0])

    for i in range(rows):
        for j in range(cols):
            if grid[i][j] == 1:
                perimeter += 4

                if i > 0 and grid[i - 1][j] == 1:      # up
                    perimeter -= 1
                if i < rows - 1 and grid[i + 1][j] == 1:  # down
                    perimeter -= 1
                if j > 0 and grid[i][j - 1] == 1:      # left
                    perimeter -= 1
                if j < cols - 1 and grid[i][j + 1] == 1:  # right
                    perimeter -= 1

    return perimeter


print(island_perimeter(
    [[0, 1, 0, 0], [1, 1, 1, 0], [0, 1, 0, 0], [1, 1, 0, 0]]))

File: JavaScript/leetcode_463.js
/**
 * @param {number[][]} grid
 * @return {number}
 */
var islandPerimeter = function (grid) {
    let perimeter = 0;
    const rows = grid.length;
    const cols = grid[0].length;
    for (let i = 0; i < rows; i++) {
        for (let j = 0; j < cols; j++) {
            if (grid[i][j] === 1) {
                perimeter += 4;
                if (i > 0 && grid[i - 1][j] === 1) perimeter--;
                if (i < rows - 1 && grid[i + 1][j] === 1) perimeter--;
                if (j > 0 && grid[i][j - 1] === 1) perimeter--;
                if (j < cols - 1 && grid[i][j + 1] === 1) perimeter--;
            }
        }
    }
    return perimeter;
};
